File: scripts/test_task_loop_preflight.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Test Task Loop Preflight
Evaluates whether current queue item is eligible for dispatch (1/2/3 complexity aware).
"""

import sys
import json
import argparse
from pathlib import Path
from datetime import datetime, timezone


def preflight(todo_path: str = None, registry_path: str = None, policy_path: str = None, lease_path: str = None, project_root: str = ".") -> dict:
    root = Path(project_root).resolve()
    
    if not todo_path:
        runtime = root / ".agents" / "task-loop"
        if not runtime.is_dir():
            runtime = root / ".codex" / "task-loop"
        todo_file = runtime / "todo.json"
        lease_file = runtime / "lease.json"
    else:
        todo_file = Path(todo_path) if Path(todo_path).is_absolute() else root / todo_path
        lease_file = Path(lease_path) if lease_path and Path(lease_path).is_absolute() else (root / lease_path if lease_path else None)

    if not todo_file.is_file():
        return {"action": "NOOP", "reason": "no_todo_items"}

    try:
        with open(todo_file, "r", encoding="utf-8") as f:
            todo = json.load(f)
    except Exception as e:
        return {"action": "INVALID", "reason": f"todo_parse_error: {e}"}

    # Check active lease
    if lease_file and lease_file.is_file():
        try:
            with open(lease_file, "r", encoding="utf-8") as f:
                lease = json.load(f)
            exp_str = lease.get("expires_at")
            if exp_str:
                exp_dt = datetime.fromisoformat(exp_str.replace("Z", "+00:00"))
                if exp_dt > datetime.now(timezone.utc):
                    return {
                        "action": "NOOP",
                        "reason": "active_lease_held",
                        "active_task": lease.get("task_id"),
                        "run_id": lease.get("run_id")
                    }
        except Exception:
            pass

    items = todo.get("items", [])
    active_items = [i for i in items if i.get("state") not in ("done", "cancelled")]
    if not active_items:
        return {"action": "NOOP", "reason": "queue_empty"}

    first_item = active_items[0]
    state = first_item.get("state")
    confirmation = first_item.get("confirmation")

    if state == "ready" and confirmation == "approved":
        goal = first_item.get("agent_goal", {})
        return {
            "action": "DISPATCH",
            "reason": "eligible",
            "item_id": first_item.get("id"),
            "complexity": goal.get("complexity", 2)
        }
    elif state in ("needs_normalization", "planning", "awaiting_user_input", "awaiting_user_confirmation"):
        return {
            "action": "DISCUSS",
            "reason": "pending_discussion",
            "item_id": first_item.get("id"),
            "state": state
        }

    return {
        "action": "NOOP",
        "reason": "item_in_progress",
        "item_id": first_item.get("id"),
        "state": state
    }


def main():
    parser = argparse.ArgumentParser(description="Test Task Loop Preflight")
    parser.add_argument("--todo", default=None, help="todo.json path")
    parser.add_argument("--registry", default=None, help="sessions.json path")
    parser.add_argument("--policy", default=None, help="policy.json path")
    parser.add_argument("--lease", default=None, help="lease.json path")
    parser.add_argument("--root", default=".", help="Project root")
    args = parser.parse_args()

    result = preflight(args.todo, args.registry, args.policy, args.lease, args.root)
    print(json.dumps(result, ensure_ascii=False))
    if result["action"] == "INVALID":
        sys.exit(2)


if __name__ == "__main__":
    main()

File: scripts/test_task_loop_preflight.js
#!/usr/bin/env node
/**
 * Test Task Loop Preflight (Node.js)
 * Evaluates whether current queue item is eligible for dispatch (1/2/3 complexity aware).
 */

const fs = require('fs');
const path = require('path');

function preflight({
  todoPath = null,
  registryPath = null,
  policyPath = null,
  leasePath = null,
  projectRoot = '.'
} = {}) {
  const root = path.resolve(projectRoot);
  let runtime = path.join(root, '.agents', 'task-loop');
  if (!fs.existsSync(runtime)) {
    runtime = path.join(root, '.codex', 'task-loop');
  }

  const todoFile = todoPath ? (path.isAbsolute(todoPath) ? todoPath : path.join(root, todoPath)) : path.join(runtime, 'todo.json');
  const leaseFile = leasePath ? (path.isAbsolute(leasePath) ? leasePath : path.join(root, leasePath)) : path.join(runtime, 'lease.json');

  if (!fs.existsSync(todoFile)) {
    return { action: 'NOOP', reason: 'no_todo_items' };
  }

  let todo;
  try {
    todo = JSON.parse(fs.readFileSync(todoFile, 'utf8'));
  } catch (e) {
    return { action: 'INVALID', reason: `todo_parse_error: ${e.message}` };
  }

  // Check active lease
  if (leaseFile && fs.existsSync(leaseFile)) {
    try {
      const lease = JSON.parse(fs.readFileSync(leaseFile, 'utf8'));
      if (lease.expires_at) {
        const expDate = new Date(lease.expires_at);
        if (expDate > new Date()) {
          return {
            action: 'NOOP',
            reason: 'active_lease_held',
            active_task: lease.task_id,
            run_id: lease.run_id
          };
        }
      }
    } catch (e) {}
  }

  const items = todo.items || [];
  const activeItems = items.filter(i => i.state !== 'done' && i.state !== 'cancelled');
  if (activeItems.length === 0) {
    return { action: 'NOOP', reason: 'queue_empty' };
  }

  const firstItem = activeItems[0];
  const state = firstItem.state;
  const confirmation = firstItem.confirmation;

  if (state === 'ready' && confirmation === 'approved') {
    const goal = firstItem.agent_goal || {};
    return {
      action: 'DISPATCH',
      reason: 'eligible',
      item_id: firstItem.id,
      complexity: goal.complexity || 2
    };
  } else if (['needs_normalization', 'planning', 'awaiting_user_input', 'awaiting_user_confirmation'].includes(state)) {
    return {
      action: 'DISCUSS',
      reason: 'pending_discussion',
      item_id: firstItem.id,
      state: state
    };
  }

  return {
    action: 'NOOP',
    reason: 'item_in_progress',
    item_id: firstItem.id,
    state: state
  };
}

function main() {
  const args = process.argv.slice(2);
  let todo = null;
  let registry = null;
  let policy = null;
  let lease = null;
  let root = '.';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--todo' && args[i + 1]) todo = args[++i];
    else if (args[i] === '--registry' && args[i + 1]) registry = args[++i];
    else if (args[i] === '--policy' && args[i + 1]) policy = args[++i];
    else if (args[i] === '--lease' && args[i + 1]) lease = args[++i];
    else if (args[i] === '--root' && args[i + 1]) root = args[++i];
    else if (!args[i].startsWith('-')) root = args[i];
  }

  const result = preflight({
    todoPath: todo,
    registryPath: registry,
    policyPath: policy,
    leasePath: lease,
    projectRoot: root
  });

  console.log(JSON.stringify(result));
  if (result.action === 'INVALID') process.exit(2);
}

if (require.main === module) {
  main();
}

module.exports = { preflight };
